src/main/java/io/oneko/search/impl/DatabaseSearchService.java

```diff
@@ -0,0 +1,48 @@
+package io.oneko.search.impl;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.google.common.collect.Streams;
+import io.oneko.event.EventDispatcher;
+import io.oneko.project.ProjectRepository;
+import io.oneko.project.event.ProjectDeletedEvent;
+import io.oneko.project.event.ProjectSavedEvent;
+import io.oneko.search.SearchResultEntry;
+import io.oneko.search.SearchService;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.springframework.stereotype.Service;
+
+@Service
+public class DatabaseSearchService implements SearchService {
+
+	private final ProjectRepository projectRepository;
+	private final Cache<String, List<SearchResultEntry>> queryResultCache = Caffeine.newBuilder()
+			.expireAfterWrite(1, TimeUnit.HOURS)
+			.build();
+
+	public DatabaseSearchService(ProjectRepository projectRepository, EventDispatcher eventDispatcher) {
+		this.projectRepository = projectRepository;
+		eventDispatcher.registerListener(event -> {
+			if (event instanceof ProjectSavedEvent) {
```

Comment:
Will new project versions or changes to a project version also trigger `ProjectSavedEvent`s?

Reply:
Yep, we're only saving Projects.